basic/23-erc865-and-erc875/Sign&Verify.md

```diff
@@ -0,0 +1,608 @@
+
+        // 3、从hash里计算出signer
+        // return recoverSigner(ethSignedMessageHash, signature);
+        // 更安全
+        return ECDSA.recover(ethSignedMessageHash, signature);
+
+    }
+    
+    // verify2直接借用了ECDSA的方法，来返回signer
+    function verify2(bytes32 _hash, bytes memory _signature) public pure returns (address) {
+        return _hash.toEthSignedMessageHash().recover(_signature);
+    }
+
+    function recoverSigner(bytes32 _ethSignedMessageHash, bytes memory _signature) public pure returns (address) {
+        (bytes32 r, bytes32 s, uint8 v) = splitSignature(_signature);
+        return ecrecover(_ethSignedMessageHash, v, r, s);
+    }
+
+		// 计算rsv
+    function splitSignature(bytes memory sig)
+        public
+        pure
+        returns (
+            bytes32 r,
+            bytes32 s,
+            uint8 v
+        )
+    {
+        require(sig.length == 65, 'invalid signature length');
+
+        assembly {
+            /*
+            First 32 bytes stores the length of the signature
+            add(sig, 32) = pointer of sig + 32
+            effectively, skips first 32 bytes of signature
+            mload(p) loads next 32 bytes starting at the memory address p into memory
+            */
+
+            // first 32 bytes, after the length prefix
+            r := mload(add(sig, 32))
+            // second 32 bytes
+            s := mload(add(sig, 64))
+            // final byte (first byte of the next 32 bytes)
+            v := byte(0, mload(add(sig, 96)))
+        }
+        // implicitly return (r, s, v)
+    }
+}
+```
+
+
+
+
+
+#### JS对签名数据计算r/s/v （做个了解）
+
+r,s,v是ecreover方法里的参数，这个是通过签名计算出来的。
+
+也可以在合约里计算，参考上面的代码。
+
+```js
+// split signature 
+// 这里的r/s/v是签名算法的偏移量，会在验证签名时用到，做个了解
+const r = signature.slice(0, 66);
+const s = "0x" + signature.slice(66, 130);
+const v = parseInt(signature.slice(130, 132), 16);
+console.log({ r, s, v });
+
+
+
+```
+
+
+
+
+
+### 二、signTypedData_v1
+
+v1基于早期的EIP-712的协议，对数据的格式要求如下：
+
+```
+{
+	  type: 'array',
+		items: {
+				type: 'object',	    
+        properties: {
+          name: {type: 'string'},
+          type: {type: 'string'}, // Solidity type as described here: https://github.com/ethereum/solidity/blob/93b1cc97022aa01e7daa9816bcc23108bbe008b5/libsolidity/ast/Types.cpp#L182
+          value: {
+            oneOf: [
+              {type: 'string'},
+              {type: 'number'},
+              {type: 'boolean'},
+            ],
+          },
+        },
+  },
+}
+```
+
+早期Web3.js里调用方式如下，现在找不到这个api去支持签名v1版本的数据了。
+
+```
+const typedData = [
+  {
+    'type': 'string',
+    'name': 'message',
+    'value': 'Hi, Alice!',
+  },
+  {
+    'type': 'uint',
+    'name': 'value',
+    'value': 42,
+  },
+];
+// 找不到了
+const signature = await web3.personal.signTypedData(typedData);
+```
+
+如果想尝试，可以用matamask提供的`eth_signTypedData`方法来唤起。效果如下：
+
+<img src="https://ipic-coda.oss-cn-beijing.aliyuncs.com/20220325/image-20220325103103575.png" alt="image-20220325103103575" style="zoom:50%;" />
```

Comment:
Image can not be loaded

Reply:
It is loaded normally for me. But I will replace it with local image